Reduce RBAC permissions for the various components

tpantelis opened the pull request:
A recent post to the **submariner-security** outlined potential vulnerabilities regarding submariner's RBAC permissions being too broad:

- create/patch/update verb of the statefulsets/replicasets resource
   - A malicious user can create a privileged container with a malicious container image capable of container escape. This would allow him/she gaining ROOT privilege of the worker node the created container deployed. Since the malicious user can control the pod scheduling parameters (e.g., replicas number, node affinity, …), he/she should be able to deploy the malicious container to every (or almost every) worker node. This means the malicious user will be able to control every (or almost every) worker node in the cluster.
   - The other permissions of statefulsets/replicasets/deployments/daemonsets also share the same risks.

In addition, many permissions aren't even used/needed. Also specifying a wildcard for resources and verbs is not best practice - Sonar [flags this](https://sonarcloud.io/project/issues?resolved=false&types=VULNERABILITY&id=submariner-io_submariner-operator&open=AY7GnmQtf1gbRJmNPGKK).

The intent of this PR is to reduce the permissions to only what's needed. See individual commits for details.

Note that for these changes to actually be applied requires updating and rebuilding **subctl**. However, I mirrored the changes in https://github.com/submariner-io/submariner-charts/pull/508 and the E2E in the **submariner-charts** repo directly tests the changes in the PR. So we can be pretty confident the changes are accurate.

Fixes https://github.com/submariner-io/submariner-operator/issues/3041

Depends on https://github.com/submariner-io/admiral/pull/905

tpantelis committed:
Reduce submariner-operator RBAC permissions

- Remove Role wildcard access to "pods" - only "get" and "list" access is
  needed which is provided by the ClusterRole.
- "services" are created for metrics so reduce Role access accordingly
- The operator does not access "endpoints", "persistentvolumeclaims",
  "events", "replicasets" or "statefulsets" so remove access.
- Reduce access to "deployments", "daemonsets" and "submariner.io" resources
  to only what's needed and remove wildcard access.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Reduce submariner-gateway RBAC permissions

- The following resources are not accessed at all so remove permissions:
  - "services"
  - "endpoints"
  - "events"
  - "configmaps"
  - "deployments"
  - "daemonsets"
  - "replicasets"
  - "statefulsets"
  - "customresourcedefinitions",
  - "operator.openshift.io/dnses"
  - "config.openshift.io/networks"
  - "monitoring.coreos.com/servicemonitors"

- The gateway pod is annotated - it only needs "get" and "update" access

- Remove wildcard access to "submariner.io" resources and specify
  exactly what's needed.

- Reduce "configmaps" access to just "get".

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Reduce submariner-routeagent RBAC permissions

- The following resources are not accessed so remove permissions:
  - "services"
  - "endpoints"
  - "events"
  - "deployments"
  - "daemonsets"
  - "replicasets"
  - "statefulsets"
  - "customresourcedefinitions",
  - "operator.openshift.io/dnses"
  - "monitoring.coreos.com/servicemonitors"

- Reduce access to "pods", "services", "secrets", "configmaps", and
  "endpoints" to "get" and "list"

- Remove wildcard access to "submariner.io" resources and specify exactly
  what's needed.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Reduce submariner-globalnet RBAC permissions

- The following resources are not accessed at all so remove permissions:
  - "services"
  - "endpoints"
  - "persistentvolumeclaims"
  - "events"
  - "configmaps"
  - "secrets"
  - "deployments"
  - "daemonsets"
  - "replicasets"
  - "statefulsets"
  - "namespaces"
  - "monitoring.coreos.com/servicemonitors"

- Reduce "pods" access to only "get", "list", "watch"

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Reduce lighthouse-agent RBAC permissions

- Remove "update" access to "services", "namespaces" and "endpoints".
- Remove wildcard access to "multicluster.x-k8s.io" resources and specify
  exactly what's needed.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Reduce lighthouse-coredns RBAC permissions

- The following resources are not accessed so remove permissions:
  - "services"
  - "endpoints"
  - "namespaces"

- Reduce "endpointslices" access to "get", "list", and "watch"

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Re-generate pkg/embeddedyamls/yamls.go

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

skitt reviewed (changes requested):
This needs some autosquashing ;-).

tpantelis committed:
Adjust code in order to remove operator RBAC permissions for namespaces

We grant namespace read permissions which is only needed to query for
the existence of the "openshift-monitoring" namespace to determine
where to create ServiceMonitors. However we can eliminate this query
and thus the permissions by trying to create ServiceMonitors in the
"openshift-monitoring" namespace and, if the error indicates the
namespace is missing, create in the provided namespace.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis commented:
I added https://github.com/submariner-io/submariner-operator/pull/3040/commits/2691550c7a23ea59f93e98275d0b609cf878834c to eliminate the operator `namespaces` permission.

Also added https://github.com/submariner-io/submariner-operator/pull/3040/commits/b268cd5e097e32387512926cc9a18b208531e2d6 and https://github.com/submariner-io/submariner-operator/pull/3040/commits/4b7e3e23d88770a6d5fa76300d31a24f97567bd6.

tpantelis committed:
Restrict config.openshift.io/networks RBAC permissions

This resource type is only used for network settings discovery to query
the "cluster" Network so restrict the RBAC to only "get" access to the
"cluster" resource name.

This applies to both the submariner-operator:

https://github.com/submariner-io/submariner/blob/85fea596f30b0e84d6962c92bb129a6b8bce8028/pkg/routeagent_driver/handlers/ovn/connection.go#L358

and route-agent components:

https://github.com/submariner-io/submariner/blob/85fea596f30b0e84d6962c92bb129a6b8bce8028/pkg/routeagent_driver/handlers/ovn/connection.go#L358

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Restrict operator ClusterRole*, ServiceAccount permissions

These were added for network-plugin syncer removal so retrict the
delete permissions to the networkplugin-syncer resource names.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

skitt commented on go.mod:
```diff
@@ -13,7 +13,7 @@ require (
 	github.com/prometheus-operator/prometheus-operator/pkg/client v0.73.2
 	github.com/prometheus/client_golang v1.19.0
-	github.com/submariner-io/admiral v0.18.0-m3
+	github.com/submariner-io/admiral v0.18.0-m3.0.20240504143439-22ee51817b45
```
This will need to be updated once https://github.com/submariner-io/admiral/pull/905 is merged.

skitt reviewed (changes requested):
Requesting changes just to make sure the Admiral import is updated to match the merged commit after https://github.com/submariner-io/admiral/pull/905 goes in.

tpantelis commented:
> Requesting changes just to make sure the Admiral import is updated to match the merged commit after [submariner-io/admiral#905](https://github.com/submariner-io/admiral/pull/905) goes in.

Done
